File: recyclerview-swipe/src/main/java/com/yanzhenjie/recyclerview/swipe/widget/DefaultItemDecoration.java
/*
 * Copyright 2017 Yan Zhenjie
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yanzhenjie.recyclerview.swipe.widget;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.View;

import com.yanzhenjie.recyclerview.swipe.SwipeMenuRecyclerView;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by YanZhenjie on 2017/8/14.
 */
public class DefaultItemDecoration extends RecyclerView.ItemDecoration {

    private Drawable mDivider;
    private int mDividerWidth;
    private int mDividerHeight;
    private List<Integer> mViewTypeList = new ArrayList<>();

    /**
     * @param color decoration line color.
     */
    public DefaultItemDecoration(@ColorInt int color) {
        this(color, 2, 2, -1);
    }

    /**
     * @param color           line color.
     * @param dividerWidth    line width.
     * @param dividerHeight   line height.
     * @param excludeViewType don't need to draw the ViewType of the item of the split line.
     */
    public DefaultItemDecoration(@ColorInt int color, int dividerWidth, int dividerHeight, int... excludeViewType) {
        mDivider = new ColorDrawable(color);
        mDividerWidth = dividerWidth;
        mDividerHeight = dividerHeight;
        for (int i : excludeViewType) {
            mViewTypeList.add(i);
        }
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        int position = parent.getChildAdapterPosition(view);
        if (position < 0) return;

        if (mViewTypeList.contains(parent.getAdapter().getItemViewType(position))) {
            outRect.set(0, 0, 0, 0);
            return;
        }

        int columnCount = getSpanCount(parent);
        int childCount = parent.getAdapter().getItemCount();

        boolean firstRaw = isFirstRaw(position, columnCount);
        boolean lastRaw = isLastRaw(position, columnCount, childCount);
        boolean firstColumn = isFirstColumn(position, columnCount);
        boolean lastColumn = isLastColumn(position, columnCount);

        if (columnCount == 1) {
            if (firstRaw) {
                outRect.set(0, 0, 0, mDividerHeight / 2);
            } else if (lastRaw) {
                outRect.set(0, mDividerHeight / 2, 0, 0);
            } else {
                outRect.set(0, mDividerHeight / 2, 0, mDividerHeight / 2);
            }
        } else {
            if (firstRaw && firstColumn) { // right, bottom
                outRect.set(0, 0, mDividerWidth / 2, mDividerHeight / 2);
            } else if (firstRaw && lastColumn) { // left, right
                outRect.set(mDividerWidth / 2, 0, 0, mDividerHeight / 2);
            } else if (firstRaw) { // left, right, bottom
                outRect.set(mDividerWidth / 2, 0, mDividerWidth / 2, mDividerHeight / 2);
            } else if (lastRaw && firstColumn) { // top, right
                outRect.set(0, mDividerHeight / 2, mDividerWidth / 2, 0);
            } else if (lastRaw && lastColumn) { // left, top
                outRect.set(mDividerWidth / 2, mDividerHeight / 2, 0, 0);
            } else if (lastRaw) { // left, top, right
                outRect.set(mDividerWidth / 2, mDividerHeight / 2, mDividerWidth / 2, 0);
            } else if (firstColumn) { // top, right, bottom
                outRect.set(0, mDividerHeight / 2, mDividerWidth / 2, mDividerHeight / 2);
            } else if (lastColumn) { // left, top, bottom
                outRect.set(mDividerWidth / 2, mDividerHeight / 2, 0, mDividerHeight / 2);
            } else { // left, bottom.
                outRect.set(mDividerWidth / 2, mDividerHeight / 2, mDividerWidth / 2, mDividerHeight / 2);
            }
        }
    }

    private int getSpanCount(RecyclerView parent) {
        RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();
        if (layoutManager instanceof GridLayoutManager) {
            return ((GridLayoutManager) layoutManager).getSpanCount();
        } else if (layoutManager instanceof StaggeredGridLayoutManager) {
            return ((StaggeredGridLayoutManager) layoutManager).getSpanCount();
        }
        return 1;
    }

    private boolean isFirstRaw(int position, int columnCount) {
        return position < columnCount;
    }

    private boolean isLastRaw(int position, int columnCount, int childCount) {
        if (columnCount == 1)
            return position + 1 == childCount;
        else {
            int lastRawItemCount = childCount % columnCount;
            int rawCount = (childCount - lastRawItemCount) / columnCount + (lastRawItemCount > 0 ? 1 : 0);

            int rawPositionJudge = (position + 1) % columnCount;
            if (rawPositionJudge == 0) {
                int rawPosition = (position + 1) / columnCount;
                return rawCount == rawPosition;
            } else {
                int rawPosition = (position + 1 - rawPositionJudge) / columnCount + 1;
                return rawCount == rawPosition;
            }
        }
    }

    private boolean isFirstColumn(int position, int columnCount) {
        if (columnCount == 1)
            return true;
        return position % columnCount == 0;
    }

    private boolean isLastColumn(int position, int columnCount) {
        if (columnCount == 1)
            return true;
        return (position + 1) % columnCount == 0;
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        drawHorizontal(c, parent);
        drawVertical(c, parent);
    }

    public void drawHorizontal(Canvas c, RecyclerView parent) {
        c.save();
        int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            int childPosition = parent.getChildAdapterPosition(child);
            if (childPosition < 0) continue;
            if (mViewTypeList.contains(parent.getAdapter().getItemViewType(childPosition))) continue;
            if (child instanceof SwipeMenuRecyclerView.LoadMoreView) continue;
            final int left = child.getLeft();
            final int top = child.getBottom();
            final int right = child.getRight();
            final int bottom = top + mDividerHeight;
            mDivider.setBounds(left, top, right, bottom);
            mDivider.draw(c);
        }
        c.restore();
    }

    public void drawVertical(Canvas c, RecyclerView parent) {
        c.save();
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            int childPosition = parent.getChildAdapterPosition(child);
            if (childPosition < 0) continue;
            if (mViewTypeList.contains(parent.getAdapter().getItemViewType(childPosition))) continue;
            if (child instanceof SwipeMenuRecyclerView.LoadMoreView) continue;
            final int left = child.getRight();
            final int top = child.getTop();
            final int right = left + mDividerWidth;
            final int bottom = child.getBottom();

            mDivider.setBounds(left, top, right, bottom);
            mDivider.draw(c);
        }
        c.restore();
    }

}

File: recyclerview-swipe/src/main/java/com/yanzhenjie/recyclerview/swipe/widget/DefaultItemDecoration.kt
/*
 * Copyright 2017 Yan Zhenjie
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yanzhenjie.recyclerview.swipe.widget

import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.support.annotation.ColorInt
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.View

import com.yanzhenjie.recyclerview.swipe.SwipeMenuRecyclerView

import java.util.ArrayList

/**
 * Default item decoration.
 *
 * Created by YanZhenjie on 2017/8/14.
 *
 * @param color           line color.
 * @param dividerWidth    line width.
 * @param dividerHeight   line height.
 * @param excludeViewType don't need to draw the ViewType of the item of the split line.
 */
open class DefaultItemDecoration @JvmOverloads constructor(
    @ColorInt color: Int,
    private val dividerWidth: Int = 2,
    private val dividerHeight: Int = 2,
    vararg excludeViewType: Int = intArrayOf(-1)) : RecyclerView.ItemDecoration() {

  private val divider: Drawable
  private val excludeTypeList = ArrayList<Int>()

  private val halfDividerWidth = dividerWidth / 2
  private val halfDividerHeight = dividerHeight / 2

  init {
    divider = ColorDrawable(color)
    excludeViewType.forEach { excludeTypeList.add(it) }
  }

  override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State?) {
    val position = parent.getChildAdapterPosition(view)

    if (position < 0) {
      return
    }

    if (excludeTypeList.contains(parent.adapter.getItemViewType(position))) {
      outRect.set(0, 0, 0, 0)
      return
    }

    val columnCount = getSpanCount(parent)
    val childCount = parent.adapter.itemCount

    val firstRaw = isFirstRaw(position, columnCount)
    val lastRaw = isLastRaw(position, columnCount, childCount)
    val firstColumn = isFirstColumn(position, columnCount)
    val lastColumn = isLastColumn(position, columnCount)

    if (columnCount == 1) {
      when {
        firstRaw -> outRect.set(0, 0, 0, halfDividerHeight)
        lastRaw -> outRect.set(0, halfDividerHeight, 0, 0)
        else -> outRect.set(0, halfDividerHeight, 0, halfDividerHeight)
      }
    } else {
      when {
        // right, bottom
        firstRaw && firstColumn -> outRect.set(0, 0, halfDividerWidth, halfDividerHeight)
        // left, right
        firstRaw && lastColumn -> outRect.set(halfDividerWidth, 0, 0, halfDividerHeight)
        // left, right, bottom
        firstRaw -> outRect.set(halfDividerWidth, 0, halfDividerWidth, halfDividerHeight)
        // top, right
        lastRaw && firstColumn -> outRect.set(0, halfDividerHeight, halfDividerWidth, 0)
        // left, top
        lastRaw && lastColumn -> outRect.set(halfDividerWidth, halfDividerHeight, 0, 0)
        // left, top, right
        lastRaw -> outRect.set(halfDividerWidth, halfDividerHeight, halfDividerWidth, 0)
        // top, right, bottom
        firstColumn -> outRect.set(0, halfDividerHeight, halfDividerWidth, halfDividerHeight)
        // left, top, bottom
        lastColumn -> outRect.set(halfDividerWidth, halfDividerHeight, 0, halfDividerHeight)
        // left, bottom.
        else -> outRect.set(halfDividerWidth, halfDividerHeight, halfDividerWidth, halfDividerHeight)
      }
    }
  }

  private fun getSpanCount(parent: RecyclerView): Int {
    val layoutManager = parent.layoutManager

    return when (layoutManager) {
      is GridLayoutManager -> layoutManager.spanCount
      is StaggeredGridLayoutManager -> layoutManager.spanCount
      else -> 1
    }
  }

  private fun isFirstRaw(position: Int, columnCount: Int): Boolean {
    return position < columnCount
  }

  private fun isLastRaw(position: Int, columnCount: Int, childCount: Int): Boolean {
    return when (columnCount) {
      1 -> position + 1 == childCount
      else -> {
        val lastRawItemCount = childCount % columnCount
        val rawCount = (childCount - lastRawItemCount) / columnCount + if (lastRawItemCount > 0) 1 else 0
        val rawPositionJudge = (position + 1) % columnCount

        val rawPosition = when (rawPositionJudge) {
          0 -> (position + 1) / columnCount
          else -> (position + 1 - rawPositionJudge) / columnCount + 1
        }
        rawCount == rawPosition
      }
    }
  }

  private fun isFirstColumn(position: Int, columnCount: Int): Boolean {
    return if (columnCount == 1) true else position % columnCount == 0
  }

  private fun isLastColumn(position: Int, columnCount: Int): Boolean {
    return if (columnCount == 1) true else (position + 1) % columnCount == 0
  }

  override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State?) {
    drawHorizontal(c, parent)
    drawVertical(c, parent)
  }

  fun drawHorizontal(canvas: Canvas, parent: RecyclerView) {
    canvas.save()
    val childCount = parent.childCount

    for (i in 0 until childCount) {
      val child = parent.getChildAt(i)
      val childPosition = parent.getChildAdapterPosition(child)

      if (childPosition < 0 || excludeTypeList.contains(parent.adapter.getItemViewType(childPosition))
          || child is SwipeMenuRecyclerView.LoadMoreView) {
        continue
      }
      val left = child.left
      val top = child.bottom
      val right = child.right
      val bottom = top + dividerHeight

      divider.setBounds(left, top, right, bottom)
      divider.draw(canvas)
    }
    canvas.restore()
  }

  fun drawVertical(canvas: Canvas, parent: RecyclerView) {
    canvas.save()
    val childCount = parent.childCount

    for (i in 0 until childCount) {
      val child = parent.getChildAt(i)
      val childPosition = parent.getChildAdapterPosition(child)

      if (childPosition < 0 || excludeTypeList.contains(parent.adapter.getItemViewType(childPosition))
          || child is SwipeMenuRecyclerView.LoadMoreView) {
        continue
      }
      val left = child.right
      val top = child.top
      val right = left + dividerWidth
      val bottom = child.bottom

      divider.setBounds(left, top, right, bottom)
      divider.draw(canvas)
    }
    canvas.restore()
  }

}
